gcache/src/gcache_rb_store.cpp

```diff
@@ -704,14 +786,116 @@ namespace gcache
            offset = -1;
         }
 
+        if (enc_encrypted != encrypt_) {
+            // if we are switching enc <-> not enc, no point in recovering
+            force_reset = true;
+        }
+
+        if (encrypt_) {
+            /* Q: Why do we store master key ID in gcache.preamble instead of
+               doing the whole key management (e.g rotation) on server side
+               and only informing Galera about the new key?
+               A: We need access to the master key when Galera is initialized
+               to be able to perform GCache recovery. At this stage storage
+               engines are not initialized yet (this is yet another question, why
+               the initialization sequence is as such).
+               If we did key management on server side, we would need to store
+               key name info somewhere. Wsrep_schema and dedicated table seems to
+               be ideal place, but again, we don't have access to it when Galera
+               starts, so another solution is keeping it in dedicated file.
+               But having yet another file on server side if we have this preamble
+               does not seem to be good idea. */
+            uint32_t crc_val = 0;
+            if (enc_crc != 0) {
+                // we've got some CRC, check if encryption data is consistent
+                gu::CRC32C crc;
+                crc.append(&enc_version, sizeof(enc_version));
+                crc.append(&enc_encrypted, sizeof(enc_encrypted));
+                crc.append(&masterKeyId_, sizeof(masterKeyId_));
+                crc.append(masterKeyUuid_.ptr(), GU_UUID_LEN);
+                crc.append(fileKey_.c_str(), fileKey_.length());
+                crc_val = crc.get();
+            }
+            if (crc_val != enc_crc) {
+                log_warn << "Encryption header CRC mismatch."
+                            << " Calculated: " << crc_val
+                            << " Expected: " << enc_crc;
+            }
+            if (enc_crc == 0 || crc_val != enc_crc) {
+                // No crc info (no header?) or crc mismatch.
+                // This will trigger new file key generation and GCache reset
+                fileKey_.clear();
+                // master key can be spoiled as well
+                masterKeyId_ = 0;
+            }
+
+            std::string mk;
+            bool allowRetry = true;
+            while (allowRetry) {
+                std::string mkName;
+                if (masterKeyId_ == 0 || masterKeyUuid_ == GU_UUID_NIL) {
+                    // no MasterKey. Generate the new one
+                    masterKeyUuid_ = gu::UUID(0,0);
+                    masterKeyId_ = 1;
+
+                    mkName = gu::CreateMasterKeyName(masterKeyUuid_, masterKeyId_);
+                    masterKeyProvider_.CreateKey(mkName);
```

Comment:
Sholudn't this error out based on the return value of CreateKey?

Reply:
this part has been reworked